Docker: docker images, microservices.

Durf1985 opened the pull request:
## In the process done:

* Assemble Docker images for a service application
* Optimize work with Docker images
* Break our application into several components
* Launch and operation of an application based on Docker images
* Launch our microservice application

## How to setup the project:

* Switch to GCP VM environment

```
eval $(docker-machine env name_your_docker_host)
```

* Build images 

```
cd your_repo/src
docker build -t <your-dockerhub-login>/ui:1.0 ./ui
docker build -t <your-dockerhub-login>/comment:1.0 ./comment
docker build -t <your-dockerhub-login>/post:1.0 ./post-py
```

* Create bridge network

```
docker network create your_name
```

* Run containers

```
docker run -d --network=reddit --network-alias=post_db --network-alias=comment_db mongo:3.2
docker run -d --network=reddit --network-alias=post <your-dockerhub-login>/post:1.0
docker run -d --network=reddit --network-alias=comment <your-dockerhub-login>/comment:1.0
docker run -d --network=reddit -p 9292:9292 <your-dockerhub-login>/ui:1.0
```

## How to check the project:

* Use the docker-machine ls command to find out the ip address of your virtual machine
* Follow the link <http://your_vm_ip_address:9292>

## PR checklist

The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-112/113 Launch app in divided containers

Added dependencies to the file requirements.txt (jinja2 MarkupSafe)
Changed image Mongo:latest to Mongo:3.2
Replaced ruby:2.2 to Ubuntu:18.04 with additional installation
of dependencies inside the image being created.

Versions of installed packages are additionally fixed inside dockerfiles

loktionovam commented on src/comment/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM ubuntu:18.04
+RUN apt-get update -qq \
+&& apt-get install -y --no-install-recommends \
```
use indentation here

loktionovam commented on src/ui/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+FROM ubuntu:18.04
+RUN apt-get update -qq \
+&& apt-get install -y --no-install-recommends \
```
use indentation

loktionovam commented on src/post-py/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM python:3.6.0-alpine
+
+WORKDIR /app
+COPY . /app
```
This code is changed frequently and it will reset your docker images cache so move it below installation of dependencies

Durf1985 replied:
In this case, this option copies the file `requirements.txt` inside the image, pip install will not work without this file.

loktionovam replied:
copy requirements.txt separately and install the dependencies first (they are changed not so frequently as application code)

Durf1985 committed:
DO-114 Reduce image size

Create 3 dockerfiles
1. Dockerfile - using multistage with alpine docker image
2. Dockerfile.alpine - a smaller image was used w/o multistage
3. Dockerfile18.04 - based on the basic ubuntu image with removal
package cache

Additionally added package and images versioning
Added tags and a non-root user to the image

loktionovam commented on src/ui/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+ENV COMMENT_SERVICE_PORT 9292
+
+FROM alpine:3.5
+ENV USER Durf
```
Parametrisation is excessive here and can lead to errors while using image

loktionovam commented on src/ui/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+COPY --from=builder $APP_HOME $APP_HOME
+WORKDIR $APP_HOME
+
+RUN adduser -D $USER
```
You can add this command above (RUN apk ...) and remove unnecessary layer
